spark/src/main/scala/filodb.spark/FiloSetup.scala

```diff
@@ -1,40 +1,89 @@
+ *
+ * Also, using a custom cluster router (NodeClusterActor), we can have a 2-node cluster system even within
+ * the same JVM.  The benefit is that local mode works exactly the same as distributed cluster mode, making
+ * the code simpler.
+ */
+trait FiloSetup extends CoordinatorSetup {
   // The global config of filodb with cassandra, columnstore, etc. sections
   def config: Config = _config.get
   var _config: Option[Config] = None
-  lazy val system = ActorSystem("filo-spark")
+  var role: String = "executor"
+
+  lazy val system = {
+    val port = systemConfig.as[Option[Int]](s"filodb.spark.$role.port").getOrElse(0)
+    ActorSystem("filo-spark", configAkka(role, port))
+  }
+
   lazy val columnStore = config.getString("store") match {
     case "cassandra" => new CassandraColumnStore(config, readEc)
     case "in-memory" => new InMemoryColumnStore(readEc)
   }
   lazy val metaStore = config.getString("store") match {
     case "cassandra" => new CassandraMetaStore(config.getConfig("cassandra"))
-    case "in-memory" => new InMemoryMetaStore
+    case "in-memory" => SingleJvmInMemoryStore.metaStore
   }
 
-  /**
-   * Initializes the config if it is not set, and start things.
-   * @param filoConfig The config within the filodb.** level.
-   */
-  def init(filoConfig: Config): Unit = _config.getOrElse {
-    _config = Some(filoConfig)
-    coordinatorActor       // Force NodeCoordinatorActor to start
-  }
+  def configAkka(role: String, akkaPort: Int): Config =
+    ConfigFactory.parseString(s"""akka.cluster.roles=[$role]
+                                 |akka.remote.netty.tcp.port=$akkaPort""".stripMargin)
+                 .withFallback(systemConfig)
+}
+
+// TODO: make the InMemoryMetaStore either distributed (using clustering to forward and distribute updates)
+// or, perhaps modify NodeCoordinator to not need metastore.
+object SingleJvmInMemoryStore {
+  import scala.concurrent.ExecutionContext.Implicits.global
+  lazy val metaStore = new InMemoryMetaStore
+}
 
-  def init(context: SparkContext): Unit = _config.getOrElse(init(configFromSpark(context)))
+object FiloDriver extends FiloSetup with StrictLogging {
+  import collection.JavaConverters._
+
+  lazy val clusterActor = system.actorOf(NodeClusterActor.props(cluster), "cluster-actor")
+  lazy val client = new ClusterClient(clusterActor, "executor", "driver")
+
+  // The init method called from a SparkContext is going to be from the driver/app.
+  def init(context: SparkContext): Unit = synchronized {
+    _config.getOrElse {
+      logger.info("Initializing FiloDriver clustering/coordination...")
+      role = "driver"
+      val filoConfig = configFromSpark(context)
+      _config = Some(filoConfig)
+      coordinatorActor
+
+      // Add in self cluster address, and join cluster ourselves
+      val selfAddr = cluster.selfAddress
+      cluster.join(selfAddr)
+      clusterActor
+      Thread sleep 1000   // Wait a little bit for cluster joining to take effect
```

Comment:
Yeah, I mean to get rid of this by having NodeClusterActor send a message back when some member is up, then from FiloSetup we can just do an actor ask.   The FiloSetup code has to be run in a Spark processing thread, so it can't be an actor itself.  I meant to investigate a better way to integrate actors into Spark processing, but no time right now.  :/      What I'm thinking is a way to run the actor receive loop inside another thread to simplify things.